docs/formatter.md

```diff
@@ -0,0 +1,648 @@
+In such cases, Black collapses the statement, and moves the comment to the end of the collapsed
+statement:
+
+```python
+# Input
+while (
+    cond1  # almost always true
+    and cond2  # almost never true
+):
+    print("Do something")
+
+# Black
+while cond1 and cond2:  # almost always true  # almost never true
+    print("Do something")
+```
+
+Ruff, like [Prettier](https://prettier.io/), expands any statement that contains trailing
+end-of-line comments. For example, Ruff would avoid collapsing the `while` test in the snippet
+above. This ensures that the comments remain close to their original position and retain their
+original intent, at the cost of retaining additional vertical space.
+
+This deviation only impacts unformatted code, in that Ruff's output should not deviate for code that
+has already been formatted by Black.
+
+<h4>Pragma comments are ignored when computing line width</h4>
+
+Pragma comments (`# type`, `# noqa`, `# pyright`, `# pylint`, etc.) are ignored when computing the width of a line.
+This prevents Ruff from moving pragma comments around, thereby modifying their meaning and behavior:
+
+See Ruff's [pragma comment handling proposal](https://github.com/astral-sh/ruff/discussions/6670)
+for details.
+
+This is similar to [Pyink](https://github.com/google/pyink) but a deviation from Black. Black avoids
+splitting any lines that contain a `# type` comment ([#997](https://github.com/psf/black/issues/997)),
+but otherwise avoids special-casing pragma comments.
+
+As Ruff expands trailing end-of-line comments, Ruff will also avoid moving pragma comments in cases
+like the following, where moving the `# noqa` to the end of the line causes it to suppress errors
+on both `first()` and `second()`:
+
+```python
+# Input
+[
+    first(),  # noqa
+    second()
+]
+
+# Black
+[first(), second()]  # noqa
+
+# Ruff
+[
+    first(),  # noqa
+    second(),
+]
+```
+
+<h4>Line width vs. line length</h4>
+
+Ruff uses the Unicode width of a line to determine if a line fits. Black's stable style uses
+character width, while Black's preview style uses Unicode width for strings ([#3445](https://github.com/psf/black/pull/3445)),
+and character width for all other tokens. Ruff's behavior is closer to Black's preview style than
+Black's stable style, although Ruff _also_ uses Unicode width for identifiers and comments.
+
+<h4>Walruses in slice expressions</h4>
+
+Black avoids inserting space around `:=` operators within slices. For example, the following adheres
+to Black stable style:
+
+```python
+# Input
+x[y:=1]
+
+# Black
+x[y:=1]
+```
+
+Ruff will instead add space around the `:=` operator:
+
+```python
+# Input
+x[y:=1]
+
+# Ruff
+x[y := 1]
+```
+
+This will likely be incorporated into Black's preview style ([#3823](https://github.com/psf/black/pull/3823)).
+
+<h4><code>global</code> and <code>nonlocal</code> names are broken across multiple lines by continuations</h4>
```

Comment:
Using `<h4>` here ensures that this long list of intentional deviations is omitted from the table of contents on the left side of the page, which IMO is good. However, it also means they can't be permalinked, which IMO is bad...

Reply:
Can you manually give them `id`? It's still not really accessible to users but at least it's possible if you know.

Reply:
I decided to move them to their own page.